File: eggfinder/game_logic.py
from collections import deque

from .board import Board
from .constants import CellType, GameConfig
from .model import GameState
from .types import ClickResult


def floodfill_reveal(board: Board, row: int, col: int) -> int:
    """BFS floodfill to reveal connected empty cells and border numbers.

    Eggs are never revealed by floodfill (matching minesweeper behavior).
    """
    queue = deque([(row, col)])
    visited = set()
    cells_revealed = 0

    while queue:
        r, c = queue.popleft()

        if (r, c) in visited:
            continue

        if not board.is_valid_position(r, c):
            continue

        if board.revealed[r][c]:
            continue

        if board.is_egg(r, c):
            continue

        visited.add((r, c))
        board.revealed[r][c] = True
        cells_revealed += 1

        cell_value = board.cells[r][c]

        if cell_value == CellType.EMPTY:
            for nr, nc in board.get_neighbors(r, c):
                if (nr, nc) not in visited:
                    queue.append((nr, nc))

    return cells_revealed


def _validate_click(board: Board, game_state: GameState, row: int, col: int) -> ClickResult | None:
    """Validate click preconditions. Returns ClickResult if invalid, None if valid."""
    if game_state.game_over:
        return ClickResult.invalid("Game is already over!", game_over=True)

    if not board.is_valid_position(row, col):
        return ClickResult.invalid(
            f"Invalid position ({row}, {col}). Must be within bounds: 0-{board.height - 1}, 0-{board.width - 1}"
        )

    if board.revealed[row][col]:
        return ClickResult.invalid(f"Cell ({row}, {col}) is already revealed!")

    return None


def _handle_egg_click(board: Board, game_state: GameState, row: int, col: int) -> ClickResult:
    """Handle clicking on an egg cell."""
    game_state.eggs_collected.add((row, col))
    board.revealed[row][col] = True
    game_state.turns_remaining += GameConfig.EGG_BONUS_TURNS
    game_state.score = calculate_score(game_state)

    game_over = game_state.turns_remaining <= 0
    if game_over:
        game_state.game_over = True

    message = f"Found an egg at ({row}, {col})! +{GameConfig.EGG_BONUS_TURNS} turns. Eggs collected: {len(game_state.eggs_collected)}/{board.egg_count}"

    return ClickResult.egg_collected(
        message=message, turns_delta=GameConfig.EGG_BONUS_TURNS, game_over=game_over
    )


def _handle_cell_click(board: Board, game_state: GameState, row: int, col: int) -> ClickResult:
    """Handle clicking on a non-egg cell."""
    cells_revealed = floodfill_reveal(board, row, col)
    game_state.turns_remaining -= 1

    game_over = game_state.turns_remaining <= 0
    if game_over:
        game_state.game_over = True

    cell_value = board.cells[row][col]
    if cell_value == CellType.EMPTY:
        message = f"Revealed {cells_revealed} cells. Turns remaining: {game_state.turns_remaining}"
    else:
        message = f"Revealed number {cell_value}. Turns remaining: {game_state.turns_remaining}"

    return ClickResult.cell_revealed(
        message=message, cells_revealed=cells_revealed, game_over=game_over
    )


def process_click(board: Board, game_state: GameState, row: int, col: int) -> ClickResult:
    """Process a player's click. Mutates board and game_state."""
    validation_result = _validate_click(board, game_state, row, col)
    if validation_result is not None:
        return validation_result

    if board.is_egg(row, col):
        return _handle_egg_click(board, game_state, row, col)
    else:
        return _handle_cell_click(board, game_state, row, col)


def check_game_over(game_state: GameState) -> bool:
    """Check if game should end."""
    return game_state.turns_remaining <= 0


def calculate_score(game_state: GameState) -> int:
    """Calculate current score based on eggs collected."""
    return len(game_state.eggs_collected)

File: eggfinder-rs/src/game_logic.rs
//! Game logic including floodfill reveal.

use std::collections::{HashSet, VecDeque};

use crate::board::Board;
use crate::constants::CellType;

/// Result of a floodfill reveal operation.
/// Contains cells with their BFS depth (distance from starting cell).
pub struct FloodfillResult {
    /// Cells revealed with their distance from the starting point: (row, col, depth)
    pub cells: Vec<(usize, usize, usize)>,
}

impl FloodfillResult {
    /// Returns the number of cells revealed.
    pub fn count(&self) -> usize {
        self.cells.len()
    }
}

/// BFS floodfill to reveal connected empty cells and border numbers.
/// Eggs are never revealed (matching minesweeper behavior).
/// Returns cells with their BFS depth for wave animation.
pub fn floodfill_reveal(board: &mut Board, row: usize, col: usize) -> FloodfillResult {
    let mut queue = VecDeque::new();
    let mut visited: HashSet<(usize, usize)> = HashSet::new();
    let mut cells = Vec::new();

    // Queue contains (row, col, depth)
    queue.push_back((row, col, 0usize));

    while let Some((r, c, depth)) = queue.pop_front() {
        if visited.contains(&(r, c)) {
            continue;
        }

        if !board.is_valid_position(r, c) {
            continue;
        }

        if board.revealed[r][c] {
            continue;
        }

        if board.is_egg(r, c) {
            continue;
        }

        visited.insert((r, c));
        board.revealed[r][c] = true;
        cells.push((r, c, depth));

        let cell_value = board.cells[r][c];

        // Only expand from empty cells (value 0)
        if cell_value == CellType::Empty.value() {
            for (nr, nc) in board.get_neighbors(r, c) {
                if !visited.contains(&(nr, nc)) {
                    queue.push_back((nr, nc, depth + 1));
                }
            }
        }
    }

    FloodfillResult { cells }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_floodfill_single_cell_with_number() {
        // Board with an egg at center - clicking corner reveals only 1 cell (a number)
        let mut board = Board::new_empty(3, 3).unwrap();
        board.add_egg(1, 1);

        let result = floodfill_reveal(&mut board, 0, 0);
        assert_eq!(result.count(), 1);
        assert!(board.revealed[0][0]);
    }

    #[test]
    fn test_floodfill_empty_board() {
        // Empty board - floodfill reveals everything
        let mut board = Board::new_empty(5, 5).unwrap();

        let result = floodfill_reveal(&mut board, 0, 0);
        assert_eq!(result.count(), 25);

        for row in &board.revealed {
            for &cell in row {
                assert!(cell);
            }
        }
    }

    #[test]
    fn test_floodfill_stops_at_numbers() {
        // Board with egg at corner - floodfill should reveal most cells
        let mut board = Board::new_empty(5, 5).unwrap();
        board.add_egg(0, 0);

        let result = floodfill_reveal(&mut board, 4, 4);

        assert!(result.count() > 1);
        assert!(board.revealed[4][4]);
    }

    #[test]
    fn test_floodfill_already_revealed() {
        let mut board = Board::new_empty(3, 3).unwrap();
        board.revealed[1][1] = true;

        let result = floodfill_reveal(&mut board, 1, 1);
        assert_eq!(result.count(), 0);
    }

    #[test]
    fn test_floodfill_does_not_reveal_eggs() {
        let mut board = Board::new_empty(5, 5).unwrap();
        board.add_egg(2, 2);

        floodfill_reveal(&mut board, 0, 0);

        // Egg should not be revealed
        assert!(!board.revealed[2][2]);
    }

    #[test]
    fn test_floodfill_reveals_numbers_but_stops() {
        // When floodfill hits a number, it reveals it but doesn't expand through it
        let mut board = Board::new_empty(5, 5).unwrap();
        board.add_egg(2, 2);

        floodfill_reveal(&mut board, 0, 0);

        // Numbers around the egg should be revealed
        assert!(board.revealed[1][1]); // This is a number (1)
        assert!(board.revealed[1][2]); // This is a number (1)

        // But egg itself should not be revealed
        assert!(!board.revealed[2][2]);
    }
}
